### PR TITLE
Create Semgrep image from apk

### PR DESCRIPTION
**Pull request by goproslowyo**
### Image Size

- [X] The Image is smaller in size than its common public counterpart.
- [ ] The Image is larger in size than its common public counterpart (please explain in the notes).

Notes:

```shell
$ docker images
REPOSITORY                      TAG       IMAGE ID       CREATED       SIZE
goproslowyo/semgrep             latest    b34fa8086cbb   3 hours ago   190MB
returntocorp/semgrep            latest    e16367d424e7   3 days ago    366MB
```

### Image Vulnerabilities

- [X] The Grype vulnerability scan returned 0 CVE(s).
- [ ] The Grype vulnerability scan returned > 0 CVE(s) (please explain in the notes).

Notes:

```shell
$ docker run --rm -v $HOME/.config/grype:/root/.config/grype -v /var/run/docker.sock:/var/run/docker.sock --name grype anchore/grype:latest goproslowyo/semgrep:latest
No vulnerabilities found

$ docker run --rm -v $HOME/.config/grype:/root/.config/grype -v /var/run/docker.sock:/var/run/docker.sock --name grype anchore/grype:latest returntocorp/semgrep  
[0024]  WARN some package(s) are missing CPEs. This may result in missing vulnerabilities. You may autogenerate these using: --add-cpes-if-none
NAME    INSTALLED  FIXED-IN  TYPE    VULNERABILITY   SEVERITY 
pip     23.1.2               python  CVE-2018-20225  High      
python  3.11.4               binary  CVE-2023-36632  High
```

### Basic Testing - K8s cluster

- [x] The container image was successfully loaded into a kind cluster.
- [ ] The container image could not be loaded into a kind cluster (please explain in the notes).

Notes:

```shell
$ kgp
NAME      READY   STATUS      RESTARTS   AGE
semgrep   0/1     Completed   0          2s
```

### Basic Testing - Package/Application

- [x] The application is accessible to the user/cluster/etc. after start-up.
- [ ] The application is not accessible to the user/cluster/etc. after start-up. (please explain in the notes).

Notes:
```shell
$ klf semgrep
Usage: semgrep [OPTIONS] COMMAND [ARGS]...

  To get started quickly, run `semgrep scan --config auto`

  Run `semgrep SUBCOMMAND --help` for more information on each subcommand

  If no subcommand is passed, will run `scan` subcommand by default

Options:
  -h, --help  Show this message and exit.

Commands:
  ci                   The recommended way to run semgrep in CI
  install-semgrep-pro  Install the Semgrep Pro Engine
  login                Obtain and save credentials for semgrep.dev
  logout               Remove locally stored credentials to semgrep.dev
  lsp                  [EXPERIMENTAL] Start the Semgrep LSP server
  publish              Upload rule to semgrep.dev
  scan                 Run semgrep rules on files
  shouldafound         Report a false negative in this project.
```

### Helm

- [ ] A Helm chart has been provided and the container image can be used with the chart.  If needed, please add a -compat package to close any gaps with the public helm chart.
- [ ] A Helm chart has been provided and the container image is not working with the chart (please explain in the notes).
- [X] A Helm chart was not provided.

Notes:
I don't think this is applicable.

### Processor Architectures

- [X] The image was built and tested for x86_64.
- [ ] The image could not be built for x86_64 (please explain in the notes).
- [ ] The image was built and tested for aarch64.
- [X] The image could not be built for aarch64. (please explain in the notes).

Notes:
I built on a local Linux workstation which has an Intel x86_64 processor.
I didn't have access to an ARM64 machine.

### Functional Testing + Documentation

- [x] Functional tests have been included and the tests are passing.  All tests have been documented in the notes section.

Notes:
Basic functional test written.

### Environment Testing + Documentation

- [x] There has not been a request and/or there is no indication that this image needs tested on a public cloud provider.
- [ ] The container image has been tested successfully on a public cloud provider (AWS, GCP, Azure).
- [x] The container image has not been tested successfully on a public cloud provider (AWS, GCP, Azure) (please explain in the notes).

Notes:
Did not test.

### Version

- [x] The package version is the latest version of the package.  The latest tag points to this version.
- [ ] The package version is the not the latest version of the package (please explain in the notes).

Notes:

### Dev Tag Availability

- [x] There is a dev tag available that includes a shell and apk tools (by depending on 'wolfi-base')
- [ ] There is not a dev tag available that includes a shell and apk tools (by depending on 'wolfi-base') (please explain in the notes).

Notes:

```shell
$ docker run --entrypoint bash -u root goproslowyo/semgrep:latest-dev -c whoami
root

$ docker run --entrypoint bash  goproslowyo/semgrep:latest-dev -c whoami
nonroot
```

### Access Control + Authentication

- [x] The image runs as `nonroot` and GID/UID are set to 65532 or upstream default
- [ ] Alternatively the username and GID/UID may be a commonly used one from the ecosystem e.g: postgres
- [ ] The image requires a non-standard username or non-standard GID/UID (please explain in the notes).

### ENTRYPOINT

- [x] applications/servers/utilities set to call main program with no arguments e.g. [redis-server]
- [ ] applications/servers/utilities not set to call main program with no arguments e.g. [redis-server] (please explain in the notes)
- [ ] base images leave empty.
- [ ] base image and not empty (please explain in the notes).
- [ ] dev variants is set to entrypoint script that falls back to system.
- [ ] dev variants is not set to entrypoint script that falls back to system (please explain in the notes).

### CMD

- [ ] For server applications give arguments to start in daemon mode (may be empty)
- [x] For utilities/tooling bring up help e.g. `–help`
- [ ] For base images with a shell, call it e.g. [/bin/sh]

### Environment Variables

- [ ] Environment variables added.
- [x] Environment variables not added and not required.

### SIGTERM

- [x] The image responds to SIGTERM (e.g., `docker kill $(docker run -d --rm cgr.dev/chainguard/nginx)`)

### Logs

- [x] Error logs write to stderr and normal logs to stdout. Logs DO NOT write to file.

### Documentation - README

- [x] A README file has been provided and it follows the README template.

### CODE REVIEW

File: images/semgrep/main.tf [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+module "tagger" {
+  source = "../../tflib/tagger"
+
+  depends_on = [module.test-latest]
+
+  tags = merge(
+    { for t in toset(concat(["latest"], module.version-tags.tag_list)) : t => module.latest.image_ref },
+    { for t in toset(concat(["latest"], module.version-tags.tag_list)) : "${t}-dev" => module.latest-dev.image_ref },
```
Comment (imjasonh) [inline images/semgrep/main.tf:52]:
```suggestion
    { for t in toset(concat(["latest-dev"], module.version-tags.tag_list)) : "${t}-dev" => module.latest-dev.image_ref },
```

**Review by imjasonh — DISMISSED** [review]
Thanks for this contribution! 🎉🎉🎉

**Review by imjasonh — DISMISSED** [review]

**Comment by imjasonh** [comment]
Hmmm, tests seem to fail with:

```
    "detail": "Test failed for ref registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190, got error: exit status 2
+ echo registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190
registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190
+ cat
+ trap 'rm -f semgrep-sample.py' EXIT
+ docker run --rm -v /home/runner/work/images/images:/work -w /work registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190 scan --lang python --config auto .
Unable to find image 'registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190' locally
registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190: Pulling from testing/semgrep
911766d2770a: Pulling fs layer
911766d2770a: Download complete
911766d2770a: Pull complete
Digest: sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190
Status: Image is up to date for registry.local:5000/testing/semgrep@sha256:e3a97d6c2518dab11fbe6c76c536fdaa0b1fb75745382fa6f3755a844cb41190
METRICS: Using configs from the Registry (like --config=p/ci) reports pseudonymous rule metrics to semgrep.dev.
To disable Registry rule metrics, use \"--metrics=off\".
Using configs only from local files (like --config=xyz.yml) does not enable metrics.

More information: https://semgrep.dev/docs/metrics

               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 3531 files tracked by git with 1091 Code rules:
                                                                                
  Language      Rules   Files          Origin      Rules                        
 ─────────────────────────────        ───────────────────                       
  <multilang>      87    9420          Community    1091                        
  terraform        99    1320                                                   
  yaml             28     716                                                   
  go               80      17                                                   
  php              33       6                                                   
  json              4       5                                                   
  bash              4       5                                                   
  c                11       4                                                   
  python          237       2                                                   
  js              179       2                                                   
  dockerfile        5       2                                                   
  ts              189       1                                                   
  java            120       1                                                   
  csharp           33       1                                                   
  html              1       1                                                   
                                                                                
missing field 'results' in JSON object of type 'CoreOutput'
Traceback (most recent call last):
  File \"/usr/lib/python3.11/site-packages/semgrep/commands/wrapper.py\", line 35, in wrapper
    func(*args, **kwargs)
  File \"/usr/lib/python3.11/site-packages/semgrep/commands/scan.py\", line 865, in scan
    ) = semgrep.run_scan.run_scan(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/run_scan.py\", line 520, in run_scan
    ) = run_rules(
        ^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/run_scan.py\", line 244, in run_rules
    ) = core_runner.invoke_semgrep_core(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/core_runner.py\", line 1194, in invoke_semgrep_core
    ) = self._run_rules_direct_to_semgrep_core(
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/core_runner.py\", line 1174, in _run_rules_direct_to_semgrep_core
    raise e
  File \"/usr/lib/python3.11/site-packages/semgrep/core_runner.py\", line 1150, in _run_rules_direct_to_semgrep_core
    return self._run_rules_direct_to_semgrep_core_helper(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/core_runner.py\", line 1085, in _run_rules_direct_to_semgrep_core_helper
    core_output = parse_core_output(output_json)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/core_output.py\", line 99, in parse_core_output
    match_results = core.CoreOutput.from_json(raw_json)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/semgrep_interfaces/semgrep_output_v1.py\", line 3483, in from_json
    results=_atd_read_list(CoreMatch.from_json)(x['results']) if 'results' in x else _atd_missing_json_field('CoreOutput', 'results'),
                                                                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File \"/usr/lib/python3.11/site-packages/semgrep/semgrep_interfaces/semgrep_output_v1.py\", line 23, in _atd_missing_json_field
    raise ValueError(f\"missing field '{json_field_name}'\"
ValueError: missing field 'results' in JSON object of type 'CoreOutput'
+ rm -f semgrep-sample.py
",
```

Does this sound like an issue with the package and/or image? Does it need some other python package in the image?

**Comment by goproslowyo** [comment]
@imjasonh I think I partially see what's happening -- `py3-semgrep.yaml` was bumped to `1.38.3` but somehow the auto-update process didn't update `semgrep.yaml` to `1.38.3` also: https://github.com/wolfi-dev/os/blob/main/semgrep.yaml#L3 because it looks like https://github.com/wolfi-dev/os/blob/main/semgrep.yaml#L59 should be `returntocorp/semgrep`. Let me make a PR to at least move forward in that direction...

**Comment by goproslowyo** [comment]
@imjasonh I think with fixing the version kerfuffle this should work properly now -- can we kick off the workflows again and see how it progresses?

**Comment by dlorenc** [comment]
> @imjasonh I think with fixing the version kerfuffle this should work properly now -- can we kick off the workflows again and see how it progresses?

Sorry, did another rebase for you!

**Review by dlorenc — APPROVED** [review]

**Pull request merged**